pennylane/templates/embeddings/qaoa.py

```diff
@@ -260,22 +259,38 @@ def circuit(weights, f=None):
         1-dimensional Ising model.
 
     """
-    wires = Wires(wires)
-    repeat = _preprocess(features, wires, weights)
-
-    if local_field == "Z":
-        local_fields = RZ
-    elif local_field == "X":
-        local_fields = RX
-    elif local_field == "Y":
-        local_fields = RY
-    else:
-        raise ValueError(f"did not recognize local field {local_field}")
 
-    for l in range(repeat):
-        # apply alternating Hamiltonians
-        qaoa_feature_encoding_hamiltonian(features, wires)
-        qaoa_ising_hamiltonian(weights[l], wires, local_fields)
+    num_params = 2
+    num_wires = AnyWires
+    par_domain = "A"
+
+    def __init__(self, features, weights, wires, local_field="Y", do_queue=True):
+
+        wires = Wires(wires)
+
+        _preprocess(features, wires, weights)
+
+        if local_field == "Z":
+            self.local_fields = RZ
+        elif local_field == "X":
+            self.local_fields = RX
+        elif local_field == "Y":
+            self.local_fields = RY
+        else:
+            raise ValueError(f"did not recognize local field {local_field}")
 
-    # repeat the feature encoding once more at the end
-    qaoa_feature_encoding_hamiltonian(features, wires)
+        super().__init__(features, weights, wires=wires, do_queue=do_queue)
+
+    def decomposition(self, features, weights, wires):
```

Comment:
Note a subtlety here: in other ops, `decomposition` is static. But since templates have a decomposition that depends on all sorts of hyperparameters, which are stored in the class attributes (since PennyLane would not know which ones to feed in here), we cannot do this here. 

I suggest making this method non-static everywhere, and removing the signature altogether - so the decomposition always depends on properties of an _instance_ of the class.

Reply:
I think we should do the same for other methods as well - `eigvals`, `matrix`, etc. It is too restrictive to assume these are static properties, in a lot of cases they depend on the operation instance.